### PR TITLE
SayHello returns hello message from peer

### PR DESCRIPTION
**Pull request by frrist**
### What
This PR lays the foundation for completing #3319 and later #3268 by modifying the `SayHello` method to receive a response. The method has also been exported so that it may be called externally to receive an updated chain head from a peer.

This is not a permanent change to the hello protocol and will likely be reverted once https://github.com/filecoin-project/go-filecoin/issues/2674 is implemented.

### CODE REVIEW

File: protocol/hello/hello.go [hunk 1]
```diff
@@ -176,7 +198,7 @@ func (hn *helloNotify) Connected(n net.Network, c net.Conn) {
 		ctx, cancel := context.WithTimeout(context.Background(), helloTimeout)
 		defer cancel()
 		p := c.RemotePeer()
-		if err := hn.hello().sayHello(ctx, p); err != nil {
+		if _, err := hn.hello().SayHello(ctx, p); err != nil {
```
Comment (frrist) [inline protocol/hello/hello.go]: Sending a hello message on this call and then again in the handler is not optimal, but it's easy, doesn't require changing the protocol bits, and doesn't rewrite much of the hello code. If this extra sending is not acceptable a proposed alternative could be to:
- modify the method `handleNewStream` such that it only sends a hello message.
- modify `SayHello` to read a message from a steam
- modify `Connect` to send a hello message, receive a response, process it and trigger the callback.

File: protocol/hello/hello.go [hunk 2]
```diff
@@ -131,34 +141,46 @@ func (h *Handler) processHelloMessage(from peer.ID, msg *Message) error {
 	return nil
 }
 
-func (h *Handler) getOurHelloMessage() *Message {
+func (h *Handler) getOurHelloMessage() (*Message, error) {
 	heaviest, err := h.getHeaviestTipSet()
 	if err != nil {
-		panic("cannot fetch chain head")
+		return nil, err
```
Comment (ZenGround0) [inline protocol/hello/hello.go]: nice

File: protocol/hello/hello.go (hunk 1 above)
Comment (ZenGround0) [reply protocol/hello/hello.go]: I'm fine with this for now.  Please put up a TODO with a follow on issue for investigating making this simpler.

Comment (frrist) [reply protocol/hello/hello.go]: implemented here: https://github.com/filecoin-project/go-filecoin/pull/3327/commits/06dbf50930399ea4c028944adf47e9c5fe8a1776, turned out to be cleaner than anticipated.

File: protocol/hello/hello.go [hunk 3]
```diff
@@ -85,31 +85,16 @@ func New(h host.Host, gen cid.Cid, helloCallback helloCallback, getHeaviestTipSe
...
-		s.Conn().Close() // nolint: errcheck
-		return
-	case ErrWrongVersion:
-		log.Debugf("code not at same version: peer has version %s, daemon has version %s, disconnecting from peer: %s", hello.CommitSha, h.commitSha, from)
-		versionErrCt.Inc(context.TODO(), 1)
-		s.Conn().Close() // nolint: errcheck
-		return
-	case nil: // ok, noop
-	default:
+	if err := cbu.NewMsgWriter(s).WriteMsg(&msg); err != nil {
```
Comment (anorth) [inline protocol/hello/hello.go]: Factor the above found lines out into a method `SendHello`.

File: protocol/hello/hello.go [hunk 4]
```diff
@@ -118,47 +103,49 @@ var ErrBadGenesis = fmt.Errorf("bad genesis block")
...
 		HeaviestTipSetCids:   heaviest.Key(),
 		HeaviestTipSetHeight: height,
 		CommitSha:            h.commitSha,
-	}
+	}, nil
 }
 
-func (h *Handler) sayHello(ctx context.Context, p peer.ID) error {
+// SayHello receives a hello message from peer `p` and returns it.
+func (h *Handler) SayHello(ctx context.Context, p peer.ID) (*Message, error) {
```
Comment (anorth) [inline protocol/hello/hello.go]: Blocking: this name is totally wrong now. Rename to `ReceiveHello`.

File: protocol/hello/hello.go [hunk 5]
```diff
@@ -171,13 +158,38 @@ func (hn *helloNotify) hello() *Handler {
...
-		p := c.RemotePeer()
-		if err := hn.hello().sayHello(ctx, p); err != nil {
-			log.Warningf("failed to send hello handshake to peer %s: %s", p, err)
+
+		// receive the hello message
+		from := c.RemotePeer()
+		hello, err := hn.hello().SayHello(ctx, from)
+		if err != nil {
+			log.Warningf("failed to receive hello handshake from peer %s: %s", from, err)
+			c.Close() //nolint: errcheck
```
Comment (anorth) [inline protocol/hello/hello.go]: Replace this idiom with `_ := c.Close()`

File: protocol/hello/hello_test.go [hunk 6]
```diff
@@ -229,3 +229,47 @@ func TestHelloMultiBlock(t *testing.T) {
...
+	mn, err := mocknet.FullMeshConnected(ctx, 2)
+	assert.NoError(t, err)
+
+	a := mn.Hosts()[0]
+	b := mn.Hosts()[1]
+
+	builder := chain.NewBuilder(t, address.Undef)
+
+	genesisTipset := builder.NewGenesis()
+	assert.Equal(t, 1, genesisTipset.Len())
```
Comment (anorth) [inline protocol/hello/hello_test.go]: You don't need to check this

Comment (frrist) [reply protocol/hello/hello_test.go]: :man_facepalming:

**Comment by whyrusleeping** [comment]
heads up: this makes go-filecoin no longer in compliance with the spec for the hello protocol.

**Comment by anorth** [comment]
@whyrusleeping I was going to write something about this in https://github.com/filecoin-project/specs/issues/478 (we're trying to work around the limitations of the defined protocol here) but upon closer inspection I decided that this was in fact in compliance. Certainly backwards compatible, but likely to be fully so.

The spec says:
> Whenever a node gets a new connection, it opens a new stream on that connection and "says hello" ... finally, closing the stream.
> ...
> Upon receiving a "hello" stream from another node, you should [read and verify]

This change just opens the hello stream more than once. A compliant node should respond every time. The spec would have to have extra words saying "and if another hello stream is opened you should reject it" in order to make this change incompatible. The spec seems to be leaving just enough room for implementers to decide when and how often to query peers.

We chose this in favour of fully disconnecting the libp2p connection with the peer, which would be more disruptive, but compatible even with the additional only-once restriction.

**Pull request merged**